### PR TITLE
Support for SAM template validation.

### PR DESCRIPTION
**Pull request by RazzM13**
A basic attempt at enabling SAM support with a reduced footprint on the current validation codebase.

Currently, the [*SAM* resource and property types specifications](https://github.com/awslabs/serverless-application-model/blob/develop/samtranslator/validator/sam_schema/schema.json) are in the *JSON-Schema* format and are incompatible with the CFN version that `cfn-lint` makes use of, which are in the *AWS Resources Specification* format.

Unfortunatelly, the above mentioned *SAM specification* is rather incomplete and therefore requires patching using the manually generated `data/sam_20161031_custom_specification.json` file, that is applied at runtime.

Compatibility with the *AWS Resources Specification* format is attained by translating the *SAM Schema* entities into corresponding *ResourcesTypes* and *PropertyTypes* via a type parameterization approach that allows *cfn-lint* to perform dynamic type and property resolution at runtime, based on the related template markup, and register logical name overrides in order to disambiguate the conflicting types during validation. 

Because the translation process requires the template markup for type resolution to take place, all intrinsic functions that alter the template must take place before the translation process can begin however, in some cases the types need to be resolved prior to intrinsic evaluation; in order to resolve the aforementioned conflict, the translation process needs to run in two phases, before the output assignment phase and before the resource property validation phase.

In addition to the resource and property types translation phase, in order to be able to validate SAM templates, *cfn-lint* must be able to reference logical resource names and related attribute values that are generated as part of the SAM transformation process threfore, based on the [SAM internals documentation](https://awslabs.github.io/serverless-application-model/internals/generated_resources.html), the aforementioned output is generated and assigned to the `SAMOutput` key of the `workingInput` variable based on the `data/sam_20161031_implicit_resources.json` file, which contains mappings of resources and their corresponding logical names that need to be generated for the specific resources that might be present in a given template.

Known limitations / considerations:
- [x] Required properties are not being validated as such (easy fix);
- [ ] [SAM policy templates](https://github.com/awslabs/serverless-application-model/blob/develop/docs/policy_templates.rst#policy-templates) are not implemented;
- [ ] Primitive types Integer, Long and Double are indistinguishable;
- [ ] Primitive types String and Timestamp are indistinguishable;
- [ ] A warning is no longer generated for when a template uses an unquoted `AWSTemplateFormatVersion`;
- [x] Logical name for generated SAM implicit types may not be correct;
- [x] Should nested parameterized types be supported? (e.g. `Map<Map<String>`);

What does work?
Sample SAM templates validate succesfully and hopefully all the old stuff still works :)

### CODE REVIEW

**Comment by RazzM13** [comment]
A basic attempt at enabling SAM support with a reduced footprint on the current validation codebase.

Currently, the [*SAM* resource and property types specifications](https://github.com/awslabs/serverless-application-model/blob/develop/samtranslator/validator/sam_schema/schema.json) are in the *JSON-Schema* format and are incompatible with the CFN version that `cfn-lint` makes use of, which are in the *AWS Resources Specification* format.

Unfortunatelly, the above mentioned *SAM specification* is rather incomplete and therefore requires patching using the manually generated `data/sam_20161031_custom_specification.json` file, that is applied at runtime.

Compatibility with the *AWS Resources Specification* format is attained by translating the *SAM Schema* entities into corresponding *ResourcesTypes* and *PropertyTypes* via a type parameterization approach that allows *cfn-lint* to perform dynamic type and property resolution at runtime, based on the related template markup, and register logical name overrides in order to disambiguate the conflicting types during validation. 

Because the translation process requires the template markup for type resolution to take place, all intrinsic functions that alter the template must take place before the translation process can begin however, in some cases the types need to be resolved prior to intrinsic evaluation; in order to resolve the aforementioned conflict, the translation process needs to run in two phases, before the output assignment phase and before the resource property validation phase.

In addition to the resource and property types translation phase, in order to be able to validate SAM templates, *cfn-lint* must be able to reference logical resource names and related attribute values that are generated as part of the SAM transformation process threfore, based on the [SAM internals documentation](https://awslabs.github.io/serverless-application-model/internals/generated_resources.html), the aforementioned output is generated and assigned to the `SAMOutput` key of the `workingInput` variable based on the `data/sam_20161031_implicit_resources.json` file, which contains mappings of resources and their corresponding logical names that need to be generated for the specific resources that might be present in a given template.

Known limitations / considerations:
- [ ] Required properties are not being validated as such (easy fix);
- [ ] [SAM policy templates](https://github.com/awslabs/serverless-application-model/blob/develop/docs/policy_templates.rst#policy-templates) are not implemented;
- [ ] Primitive types Integer, Long and Double are indistinguishable;
- [ ] Primitive types String and Timestamp are indistinguishable;
- [ ] A warning is no longer generated for when a template uses an unquoted `AWSTemplateFormatVersion`;
- [ ] Logical name for generated SAM implicit types may not be correct;
- [ ] Should nested parameterized types be supported? (e.g. `Map<Map<String>`);

What does work?
Sample SAM templates validate succesfully and hopefully all the old stuff still works :)

**Comment by martysweet** [comment]
Are we doing the transformation? From your summary (nice and in-depth, thanks), I assume we are?

I don't think this is necessary, I was more thinking of the SAM resources being treated merely an extension of the current resource specification. Is it not safe to assume that a valid SAM will generate a valid unravelled CFN template? Additionally, if we transform then give messages, the error feedback won't be great. Let me know if this no sense!

Will continue to look through...

**Comment by RazzM13** [comment]
Hi @martysweet, in regards to:

> Are we doing the transformation? 

Yes, you are correct in assuming that we're doing the [SAM transformation](https://github.com/RazzM13/cfn-lint/blob/0b7aca79669944d5e818d4263d66b5e1ed44397b/src/validator.ts#L367-L421) however it is just a shallow version of it that generates skelleton resource instances (i.e. no `Properties` section) and assigns the output to the `SAMOutput` key of the `workingInput` variable under their corresponding logical name keys, which have been derived from the [SAM internals documentation](https://awslabs.github.io/serverless-application-model/internals/generated_resources.html). This is so that the aforementioned skeleton instances can be referenced in `Ref` and `GetAtt` intrinsics that may occur within template resource property values or even in other sections such as `Outputs`, as per [here](https://github.com/RazzM13/cfn-lint/blob/0b7aca79669944d5e818d4263d66b5e1ed44397b/testData/valid/yaml/sam_20161031_lambda_edge.yaml#L28) and respectively [here](https://github.com/RazzM13/cfn-lint/blob/0b7aca79669944d5e818d4263d66b5e1ed44397b/testData/valid/yaml/sam_20161031_lambda_edge.yaml#L75); the `LambdaEdgeFunctionSample.Version` referenced in the sample is actually a SAM generated `AWS::Lambda::Version` resource, explicitly stated [here](https://github.com/awslabs/serverless-application-model/blob/develop/versions/2016-10-31.md#referencing-lambda-version--alias-resources).

*PS*: My interpretation of the [SAM internals documentation](https://awslabs.github.io/serverless-application-model/internals/generated_resources.html) logical name formatting is very likely to be incorrect, it would be terrific if you could have a look over it yourself so that we can shed some light upon this matter.

> Is it not safe to assume that a valid SAM will generate a valid unravelled CFN template?

I concur however, how do we know if it's a valid SAM when a property value references a resource-to-be via the logical name? The logical name may be incorrect or in the case of `GetAtt` the attribute name may be incorrect or even the value in the current context may be invalid...

> Additionally, if we transform then give messages, the error feedback won't be great.

I'm not really sure about this, AFAICT the errors should actually be in the real template and not in the `SAMOutput` section, though I haven't tested for this.

> I was more thinking of the SAM resources being treated merely an extension of the current resource specification.

Initially that's what I thought too but as I delved deeper, the fallacy became apparent and glympses of what used to be an easy fix were now shadowed by a hazy madness full of uncertainty that encompasses the 7 circles of SAM. This led to the parameterized type approach, followed by the schema patch and last but not least SAM fulfillment.

> I don't think this is necessary

I'm open to ideas, perhaps one would be to just skip validation for items that reference any logical names that are not part of the template or that match the name of a `AWS::Serverless::*` resource, followed by a `.` and anything else (e.g. `LambdaEdgeFunctionSample.Version`).

> Will continue to look through...

:+1:

File: data/sam_20161031_custom_specification.json [hunk 1]
```diff
@@ -0,0 +1,230 @@
+{
```
Comment (martysweet) [inline data/sam_20161031_custom_specification.json]: Did you make this manually? Is it convertible from something else?

File: src/resourcesSpec.ts [hunk 2]
```diff
@@ -173,8 +389,6 @@ export function getItemType(baseType: string, propType: string, key: string) {
 
     if (!property.ItemType) {
         return undefined;
-    } else if (property.ItemType === 'Tag') {
```
Comment (martysweet) [inline src/resourcesSpec.ts]: Are there tests covering the 'Tag' type? I seem to remember it not being specified correctly in the resource spec from aws.

File: src/test/validatorTest.ts [hunk 3]
```diff
@@ -645,14 +645,15 @@ describe('validator', () => {
...
 
-        it('1 unquouted template format version should return an object with validTemplate = true, no crit errors, 1 warn error', () => {
-            const input = 'testData/valid/yaml/valid_unquoted_template_version.yaml';
-            let result = validator.validateFile(input);
-            expect(result).to.have.deep.property('templateValid', true);
-            expect(result['errors']['crit']).to.have.lengthOf(0);
-            expect(result['errors']['warn']).to.have.lengthOf(1);
-            expect(result['errors']['warn'][0]['message']).to.contain('AWSTemplateFormatVersion is recommended to be of type string');
-        });
+        // TODO: is this really needed?
```
Comment (martysweet) [inline src/test/validatorTest.ts]: I can't remember, prehaps a git blame will shed more light on this test

File: src/test/validatorTest.ts [hunk 4]
```diff
@@ -704,7 +705,7 @@ describe('validator', () => {
             let result = validator.validateFile(input);
             expect(result).to.have.deep.property('templateValid', false);
             expect(result['errors']['crit']).to.have.lengthOf(1);
-            expect(result['errors']['crit'][0]['message']).to.contain('Required property Key missing for type Tag');
+            expect(result['errors']['crit'][0]['message']).to.contain('Required property Key missing for type AWS::EC2::Instance.Tag');
```
Comment (martysweet) [inline src/test/validatorTest.ts]: What about other resource tags on things like DynamoDB tables and Lambda functions?

File: src/test/validatorTest.ts [hunk 5]
```diff
@@ -1146,6 +1147,7 @@ describe('validator', () => {
 
         it('should be able to determine the type of every property of every propertytype', () => {
             for (let propertyTypeName in awsResources.PropertyTypes) {
+                if (propertyTypeName == 'Tag') { continue; }
```
Comment (martysweet) [inline src/test/validatorTest.ts]: ?

File: src/validator.ts [hunk 6]
```diff
@@ -234,6 +217,290 @@ function validateWorkingInput(passedOptions?: Partial<ValidateOptions>) {
...
 
+function applySpecificationOverrides() {
+
+    // extend resources specification with SAM specifics
+    if (!!~workingInputTransform.indexOf('AWS::Serverless-2016-10-31')) {
+        resourcesSpec.extendSpecification(samData.samResources20161031);
+    }
+
+    // normalize Tag property type access
+    let tagPropertyTypes: any = {};
```
Comment (martysweet) [inline src/validator.ts]: Why is this required?

File: data/sam_20161031_custom_specification.json (hunk 1 above)
Comment (RazzM13) [reply data/sam_20161031_custom_specification.json]: Unfortunatelly yes, it is an extension of the SAM specification due to the fact that certain properties (e.g. `AutoPublishAlias` of `AWS::Serverless::Function`) are missing from the JSON schema file but are present in the documentation; some properties such as `MethodSettings` of `AWS::Serverless::Api` accept as value a CFN property-type, in this case `AWS::ApiGateway::Deployment.MethodSetting`, which for the sake of convenience I've chosen to simply clone as `AWS::Serverless::Api.MethodSetting` and therefore not alter the current way `cfn-lint` looks up property types though, perhaps it might be cleaner but certainly more convoluted to do the opposite... if you have any idea on how to tackle this or perhaps even avoid it completely then please feel free to share :)

File: src/resourcesSpec.ts (hunk 2 above)
Comment (RazzM13) [reply src/resourcesSpec.ts]: Yes, the way the `Tag` property type is defined in the AWS Resource Specification is inconsistent with the rest.

If I remember correctly, I decided to remove the artificial way of looking up `Tag` property-types in favor of augmenting the specification at runtime by adding a `ResourceType.Tag` property type for each of the defined resources, this approach seemed more consistent / cleaner to me and also helped to avoid an endless loop that was occuring when `isParameterizedTypeFormat` was being called from somewhere (I can't recall now).

File: src/test/validatorTest.ts (hunk 5 above)
Comment (RazzM13) [reply src/test/validatorTest.ts]: This test skips checking the `Tag` property type as it's no longer needed nor do I believe it makes sense anymore as the correct reference should be `ResourceType.Tag` and those are being checked.

File: src/validator.ts (hunk 6 above)
Comment (RazzM13) [reply src/validator.ts]: This actually creates the specific `ResourceType.Tag` property types, as depicted above.

File: src/test/validatorTest.ts (hunk 3 above)
Comment (RazzM13) [reply src/test/validatorTest.ts]: I've just tried doing just that but I haven't been able to come up with some definitive answer, this either means that my git foo isn't strong enough (highly likely) or that perhaps it was added in some mass refactoring ... in essence, being it's just a warning, I'm not really sure that this is critical nor do I know of an easy fix for this so I leave that up to you ...

**Comment by martysweet** [comment]
Hi @RazzM13, 

Do you need any further input from me to continue work on this?

Most appreciated,
Marty

**Comment by RazzM13** [comment]
Hi @martysweet,
Well, since my last reply, I was actually waiting on a bit of feedback from you (hence the long delay on my part :)), what I'm mostly interested in knowing is if you're ok with the approach as whole, especially the current limitations/considerations that I've mentioned in the above list; I would advise that, for the time being, we just accept the constraints imposed by items 2 - 5 and perhaps 7, and focus on 1 and most importantly 6.
In regards to consideration item 6 `Logical name for generated SAM implicit types may not be correct` could you perhaps have a look over the  [SAM internals documentation](https://awslabs.github.io/serverless-application-model/internals/generated_resources.html) and help me figure out what the correct resource naming scheme should be, as my past interpretation was conflicting with a sample SAM template? ... perhaps the best way to tackle this would be to acquire a few real world valid templates and see exactly how the resources are named.

File: data/sam_20161031_implicit_resources.json [hunk 7]
```diff
@@ -0,0 +1,49 @@
+{
+  "AWS::Serverless::Function": {
+    "AWS::Lambda::Function": "<LogicalName>",
+    "AWS::IAM::Role": "<LogicalName>Role"
+  },
+  "AWS::Serverless::Function#AutoPublishAlias": {
+    "AWS::Lambda::Version": "<LogicalName>.Version",
```
Comment (martysweet) [inline data/sam_20161031_implicit_resources.json]: Shouldn't this be:

```
<LogicalName>Version<SHA>
```

File: data/sam_20161031_implicit_resources.json [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    "AWS::Lambda::Alias": "<LogicalName>Aliaslive"
+  },
+  "AWS::Serverless::Function#DeploymentPreference": {
+    "AWS::CodeDeploy::Application": "ServerlessDeploymentApplication",
```
Comment (martysweet) [inline data/sam_20161031_implicit_resources.json]: Need to ensure this is only ever generated once

File: data/sam_20161031_implicit_resources.json [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    "AWS::CodeDeploy::DeploymentGroup": "<LogicalName>DeploymentGroup",
+    "AWS::IAM::Role": "CodeDeployServiceRole"
+  },
+  "AWS::Serverless::Function#Events#ItemId<ApiEvent>": {
+    "AWS::ApiGateway::RestApi": "ServerlessRestApi",
+    "AWS::ApiGateway::Stage": "ServerlessRestApiProdStage",
+    "AWS::ApiGateway::Deployment": "ServerlessRestApIdeployment<SHA>",
```
Comment (martysweet) [inline data/sam_20161031_implicit_resources.json]: `ServerlessRestApiDeployment<SHA>`

File: data/sam_20161031_implicit_resources.json [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    "AWS::Lambda::Permission": "<LogicalName><ItemId>PermissionProd"
+  },
+  "AWS::Serverless::Function#Events#ItemId<S3Event>": {
+    "AWS::Lambda::Permission": "<LogicalName><ItemId>Permission"
```
Comment (martysweet) [inline data/sam_20161031_implicit_resources.json]: > Existing MyBucket resource is modified to append NotificationConfiguration property where the Lambda function trigger is defined

File: data/sam_20161031_implicit_resources.json [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    "AWS::Lambda::Permission": "<LogicalName><ItemId>Permission",
+    "AWS::Events::Rule": "<LogicalName><ItemId>"
+  },
+  "AWS::Serverless::Function#Events#ItemId<CloudWatchEventEvent>": {
+    "AWS::Lambda::Permission": "<LogicalName><ItemId>Permission",
+    "AWS::Events::Rule": "<LogicalName><ItemId>"
+  },
+  "AWS::Serverless::Api": {
+    "AWS::ApiGateway::RestApi": "<LogicalName>",
+    "AWS::ApiGateway::Stage": "<LogicalName><ItemId>Stage",
```
Comment (martysweet) [inline data/sam_20161031_implicit_resources.json]: Can `ItemId` be used here?

**Comment by martysweet** [comment]
I am happy with the approach. It might be worth taking a look at how this new kid of the block supports SAM validation https://github.com/awslabs/cfn-python-lint, for any simplifications. :wink: 

1 needs to be done
2,3,4,5 are not too important
6 sorry for the SPAM, should have started a review :+1: Throw me an example if you still aren't happy with it, for the most part, I agree with what you have.

**Comment by RazzM13** [comment]
Hi @martysweet, sorry for the long delay, I've just managed to get my hands back on this again and, as per the last comments, I've added support for validating required properties and finally managed to  figure out and sort of fix what was happening with the naming/referencing of the resources resulting from the *SAM* transform. 

In contrast to *CFN* (AFAICT),  *SAM* can reference runtime generated resources by referencing a property of the source/parent *AWS::Serverless::\** resource that points to the resource to-be (e.g. referencing `MyLambdaFunction.Version` will return the corresponding `AWS::Lambda::Version` generated by SAM) therefore I've attempted to support this feature by adding an extra property called `refProperty` to the transform definitions of the `sam_20161031_implicit_resources.json` file that upon creating a specific SAM resource will create a property on the resource that triggered the creation of the aforementioned and that links to the newly created resource, named according to the `refProperty` value.

In regards to the Logical ID naming, I believe the templates of the `sam_20161031_implicit_resources.json` file should now be OK however, `cfn-lint` will not be able to correctly generate the naming for the `AWS::ApiGateway::Deployment` resource that arises when an `ApiEvent` event is present under the `Events` section of the `AWS::Serverless::Function` resource definition due to the fact that we do not have access to the contents of the *Swagger* definition that gets generated by SAM and therefore we cannot establish the correct hash ID; also, the aforementioned is applicable for resources that depend on the `CodeUri` and `DefinitionUri` properties when they reference a local file that should be converted to an S3 location by `aws cloudformation package` or something similar. Perhaps we may need to do things differently here? some real-world templates might help us to figure out to best way to tackle this. 

Last but not least, please let me know your thoughts and how you would like us to proceed with it.

**Comment by martysweet** [comment]
Looking good @RazzM13. 

>Perhaps we may need to do things differently here? 
Ah that's tricky, so we are stuck because SAM transforms the template before uploading? Perhaps on these properties, we can raise a notice to say, this property should be an S3 url, for example.

Is there much more to do on this until you are happy for it to be in a major release? Once you are happy, I will spend a few hours going through it all in more detail (as it's quite large).

Let me know if you need anything in the meantime.

**Comment by RazzM13** [comment]
Hi @martysweet, I've just managed to rebase this on master and following, one test started to fail due to the fact that it is expecting the `Key` property of the `AWS::EC2::Instance.Tag` property type to be required, when in fact it is not however, it used to be required prior to commit c594aa3be32db896bb3823a4fa87b9e34dbc09a4; I am guessing here that this is actually an error on behalf of AWS and it should actually be required...
PS: as a side note, CircleCI reported the build to be successful however, if you look at the execution output of the `test` phase of the build status then you can clearly see that this is not the case... perhaps the return code of the test command is being overridden somehow???

**Comment by martysweet** [comment]
Yes it should definitely be required. 

I can't seem to find the ItemType of tag in that commit https://raw.githubusercontent.com/martysweet/cfn-lint/c594aa3be32db896bb3823a4fa87b9e34dbc09a4/data/aws_resources_specification.json, doesn't appear to be in the latest either https://d1uauaxba7bl26.cloudfront.net/latest/gzip/CloudFormationResourceSpecification.json 👍 

Good spot on the Circle issue! Will look into it

**Comment by RazzM13** [comment]
Hi @martysweet. just a short update on what I'm currently facing / working on:
1) there is a lot of new functionality related to type inference (for de-parameterization purposes) and no tests, I'm currently working on refactoring this to use the existing code (e.g. isJson, isBoolean, etc.) and adding tests as I go along;
2) found a type inference bug when attempting to validate a real-world template, I'll be looking into this, maybe it will just go away with the "refactoring" :) ;
3) apparently a lot of real-world templates do not provide the `CodeUri` and `Runtime` properties and I'm guessing they're relying upon `aws cloudformation package` to upload the local artifacts and fill those in before submitting the template to CF (though I have absolutely no idea how `aws cloudformation package` figures out that it should upload the local directory, without it being referenced in the `CodeUri`, as depicted in their docs...).

PS: Just noticed that some tests are failing (timed out) as part of the latest changes to master, I'll have a look at them later on... Hope everything is ok on your end :)

**Comment by RazzM13** [comment]
Hey @martysweet, I've just finished adding all the major functionality needed for validating SAM templates and I've done some fixes / refactoring / small optimizations along the way and I believe we're now at a stage that we can push for public release under the umbrella term `experimental support` and wait for the bugs to rain in :)

One of the major changes I've done is added support for validating SAM Globals that I believe are being used a lot in real-world templates and changed to a manually generated SAM CFN specification file `data/sam_20161031_cfn.json` that should help with debugging future SAM related issues as well as my mistaken belief that it would have shortened the test runtime.

Last but not least, regarding maintenance, I feel the need to highlight that when doing releases you will have to keep an eye on the latest version of `https://github.com/awslabs/serverless-application-model/blob/develop/samtranslator/validator/sam_schema/schema.json` and propagate any changes to our local file `data/sam_20161031_schema.json` and if there are no additional overrides to be declared in `data/sam_20161031_custom_specification.json`, you can just run the `sam2CFN` utility that will generate the final SAM CFN specification file `data/sam_20161031_cfn.json` and upon committing that to master finally, that should be it.

Please let me know your thoughts on *

**Comment by martysweet** [comment]
Thanks for all the work @RazzM13. I will try and go through it all over the next couple of days and will do a release.

Will using `sam2CFN` do all the translation required?

**Comment by RazzM13** [comment]
Your welcome and thank you for providing all of us with such as nice project.
Please have a look at it all and if you need any clarifications or would like to do things differently don't hesitate to ask :)
Last but not least, yes, `sam2CFN` will generate the final `data/sam_20161031_cfn.json` that will be used by `cfn-lint`. It does so by firstly converting the SAM schema (i.e. `data/sam_20161031_schema.json`) to CFN and then does a merge of the generated schema with the contents of the `data/sam_20161031_custom_specification.json` file where, conflicting definitions will be overridden in the former by the latter.

File: src/resourcesSpec.ts [hunk 12]
```diff
@@ -82,36 +94,274 @@ export function getResourceTypeAttribute(type: string, attributeName: string): A
     return resourceAttribute
 }
 
-function getPropertyType(type: string){
-    const propertyType =  specification.PropertyTypes[type];
-    if (!propertyType) {
-        throw new NoSuchPropertyType(type);
+function getPropertyType(type: string): awsData.ResourcePropertyType {
+
+    // destructure property name
```
Comment (martysweet) [inline src/resourcesSpec.ts]: could we have a comment with the examples of what is expected here? The reuse of variables isn't the clearest. 

Something like
`// ex. baseType = AWS::EC2::Instance, baseTypeName = ABCD, etc`

Will help someone (and me) understand the code easier

File: README.md [hunk 13]
```diff
@@ -112,6 +112,7 @@ Template invalid!
 * Go to the documentation from Command Line (see above examples)
 * Detecting invalid property types
 * Detect missing required properties
+* SAM template validation (experimental support)
```
[the remark this answers is not in the document]
Comment (martysweet) [reply README.md]: also a quick paragraph on how the SAM stuff works would be great (with the technical words used in the code)

File: src/resourcesSpec.ts [hunk 14]
```diff
@@ -149,32 +399,39 @@ export function isAdditionalPropertiesEnabled(resourceType: string){
...
 
 export function getItemType(baseType: string, propType: string, key: string) {
     const property = getProperty(propType, key);
-
     if (!property.ItemType) {
         return undefined;
-    } else if (property.ItemType === 'Tag') {
-        return 'Tag'
+    } else if (isComplexType(property.ItemType)) {
+        return property.ItemType;
```
Comment (martysweet) [inline src/resourcesSpec.ts]: Nice

File: src/resourcesSpec.ts [hunk 15]
```diff
@@ -211,3 +468,74 @@ export function getRequiredProperties(type: string){
 export function extendSpecification(spec: any){
     Object.assign(specification, mergeOptions(specification, spec));
 }
+
+/**
+ * Allows overriding definitions based on logical name.
+ * Subsequent registrations DO clobber prior ones.
```
Comment (martysweet) [inline src/resourcesSpec.ts]: Better word than clobber? Overwrite?

File: src/resourcesSpec.ts [hunk 16]
```diff
@@ -211,3 +468,74 @@ export function getRequiredProperties(type: string){
...
+    // override
+    extendSpecification({
+        [typeSection]: {[name]: mergeOptions(oldSpec, spec)}
+    });
+
+}
+
+/**
+ * Allows overriding definitions based on type.
+ * Subsequent registrations DO clobber prior ones.
```
Comment (martysweet) [inline src/resourcesSpec.ts]: Again with clobber

File: src/sam2CFN.ts [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+      processDefinition(type, samSchemaDef, awsResourcesSpec);
+    }
+    // enhance with custom specification
+    awsResourcesSpec = mergeOptions(awsResourcesSpec, customSpecification);
+
+    return awsResourcesSpec;
+};
+
+
+program
```
Comment (martysweet) [inline src/sam2CFN.ts]: Do we have any tests for this?

File: src/test/validatorTest.ts (hunk 3 above)
Comment (martysweet) [reply src/test/validatorTest.ts]: Lets remove this

File: src/test/validatorTest.ts [hunk 18]
```diff
@@ -1400,4 +1408,333 @@ describe('validator', () => {
...
+
+          it('should be able to infer a String value', () => {
+              let input = 'somethingBeautiful';
+              let result = validator.__TESTING__.inferPrimitiveValueType(input);
+              expect(result).to.equal('String');
+          });
+
+          it('should be able to infer a Timestamp value', () => {
+              let input = '2012-12-30T20:12:22.222222-08:00';
+              let result = validator.__TESTING__.inferPrimitiveValueType(input);
```
Comment (martysweet) [inline src/test/validatorTest.ts]: Might be worth testing with more timestamp ISO8601 types here if we see it being used?

For ex. no microseconds, and a zulu timezone etc

File: src/validator.ts [hunk 19]
```diff
@@ -209,19 +179,34 @@ function validateWorkingInput(passedOptions?: Partial<ValidateOptions>) {
...
+    // Apply template overrides
+    applyTemplateOverrides();
 
     // Check parameters and assign outputs
     assignParametersOutput(options.guessParameters);
 
     // Evaluate Conditions
     assignConditionsOutputs();
 
-    // Apply resource specification overrides
```
Comment (martysweet) [inline src/validator.ts]: Can this comment be merged into where this is now applied?

File: src/validator.ts [hunk 20]
```diff
@@ -245,6 +233,524 @@ function validateWorkingInput(passedOptions?: Partial<ValidateOptions>) {
...
+        if (!!workingInputGlobals && (typeof workingInputGlobals == 'object')) {
+            for (let resourceType of Object.keys(workingInputGlobals)) {
+                let resourceTemplate = workingInputGlobals[resourceType];
+                if (!!resourceTemplate) {
+                    let targetType: string | undefined;
+                    let allowedProperties: string[] = [];
+                    switch(resourceType) {
+                        case 'Function':
+                            targetType = 'AWS::Serverless::Function';
+                            allowedProperties = [
```
Comment (martysweet) [inline src/validator.ts]: Can these be in an external JSON?

**Comment by martysweet** [comment]
@RazzM13 Let me know if you have any questions/need help

**Comment by RazzM13** [comment]
Hi @martysweet, I'm currently working on the latest changes, the tests are taking a bit more effort but my lack of time is actually to blame here, hopefully by the end of tomorrow I'll manage to commit the changes. 
Hope that's ok and sorry for the delay again :)

**Comment by martysweet** [comment]
No worries! No rush :)

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (RazzM13) [reply README.md]: Everything should now be in the README file, please check it out.

File: src/resourcesSpec.ts (hunk 12 above)
Comment (RazzM13) [reply src/resourcesSpec.ts]: Hopefully, it should be OK now...

File: src/sam2CFN.ts (hunk 17 above)
Comment (RazzM13) [reply src/sam2CFN.ts]: For testing purposes, I've had to split this into two components a CLI part, namely `sam2CFN.ts `, and a function library, namely `sam2CFNUtils.ts `. Please let me know if this is OK and/or you would like any additional tests added.

File: src/test/validatorTest.ts (hunk 18 above)
Comment (RazzM13) [reply src/test/validatorTest.ts]: I'm not really sure because I've been mainly targeting the integration that occurs within the `inferPrimitiveValueType` function and relied upon the functional tests for each specific function, such as `isTimestamp`, to be exhaustive. 
Please let me know if you think it should be done differently.

File: src/validator.ts (hunk 19 above)
Comment (RazzM13) [reply src/validator.ts]: Does this still apply? ...

**Comment by RazzM13** [comment]
Hey @martysweet, just managed to push the latest changes, please have a look through and let me know if anything is missing or incorrect or otherwise :)

File: src/sam2CFN.ts (hunk 17 above)
Comment (martysweet) [reply src/sam2CFN.ts]: Is the Utils tested for sanity?

**Comment by martysweet** [comment]
LGTM! Will go over the README tomorrow (Saturday) and merge into master.

File: src/sam2CFN.ts (hunk 17 above)
Comment (RazzM13) [reply src/sam2CFN.ts]: By *Utils* I believe you are referring to `sam2CFNUtils.ts` and if so, then please define *sanity* ... 
Basically both have tests, the tests regarding the library part should cover all the functionality used in the conversion process and the CLI has a unit test that validates that the parameters are actually being parsed and that the default values aren't used if the parameters have actually been specified on the command-line.
I was considering adding a test for the CLI that actually touched a file on the FS but I thought that perhaps it might be best to just mock the FS functions and check if they are being called with the correct arguments.
Anyway, just let me know how you would like to proceed with this.

Comment (martysweet) [reply src/sam2CFN.ts]: Ok, that's great thanks! Just wanted to check the functionality of the library part was fully covered.
